listenbrainz/listenstore/timescale_utils.py

```diff
@@ -1,41 +1,245 @@
+            -- cannot use = min(dl.listened_at). see note 3
+            HAVING min_listened_at = ANY(array_agg(dl.listened_at))
+        ), calculate_new_ts AS (
+            SELECT user_id
+                 , new_ts.min_listened_ts AS new_listened_at
+              FROM update_ts u
+              -- for each user calculate the new minimum timestamp
+              JOIN LATERAL (
+                    SELECT min(listened_at) AS min_listened_ts
+                      FROM listen l
+                        -- new minimum will be greater than the last one
+                     WHERE l.listened_at >= u.min_listened_at
+                       AND l.user_id = u.user_id
+                       AND l.created <= u.created
+                   ) AS new_ts
+                ON TRUE
+        )
+            UPDATE listen_user_metadata lm
+               SET min_listened_at = new_listened_at
+              FROM calculate_new_ts mt
+             WHERE lm.user_id = mt.user_id
+    """
+
+    # check for which users the listen of maximum listened_at was deleted, for those users
+    # calculate new maximum listened_at
+    update_listen_max_ts = """
+        WITH update_ts AS (
+            SELECT user_id, max_listened_at, lm.created
+              FROM listen_delete_metadata dl
+              JOIN listen_user_metadata lm
+             USING (user_id)
+             WHERE dl.id <= :max_id
+          GROUP BY user_id, max_listened_at, lm.created
+            -- check if max_listened_at is in list of listened_at of deleted listens,
+            -- cannot use = max(dl.listened_at). see note 3
+            HAVING max_listened_at = ANY(array_agg(dl.listened_at))
+        ), calculate_new_ts AS (
+            SELECT user_id
+                 , new_ts.max_listened_ts AS new_listened_at
+              FROM update_ts u
+                -- for each user calculate the new maximum timestamp
+              JOIN LATERAL (
+                    SELECT max(listened_at) AS max_listened_ts
+                      FROM listen l
+                        -- new maximum will be lesser than the last one
+                     WHERE l.listened_at <= u.max_listened_at
+                       AND l.user_id = u.user_id
+                       AND l.created <= u.created
+                   ) AS new_ts
+                ON TRUE
+        )
+            UPDATE listen_user_metadata lm
+               SET max_listened_at = new_listened_at
+              FROM calculate_new_ts mt
+             WHERE lm.user_id = mt.user_id
+    """
+    delete_user_metadata = "DELETE FROM listen_delete_metadata WHERE id <= :max_id"
+
+    with timescale.engine.begin() as connection:
+        result = connection.execute(select_max_id)
+        row = result.fetchone()
+
+        if not row:
+            logger.info("No pending deletes")
+            return
+
+        max_id = row["max_id"]
+        logger.info("Found max id in listen_delete_metadata table: %d", max_id)
```

Comment:
Is there any value to leaving these info statements?

Reply:
Yes, I think it may come useful if something goes wrong during the cron job run. At least in tracking how much time each query is taking to execute.